### ISSUE TITLE
words.protocols.jabber.sasl uses base64.b64encode, which breaks compatibility with Python 2.3.x

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2461]
|<img alt="murphy_slaw's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| murphy_slaw reported|
|-|-|
|Trac ID|trac#2461|
|Type|defect|
|Created|2007-02-19 21:34:36Z|

Twisted 2.5.0 release notes say:

"Twisted is an event-based framework for internet applications which
works on Python 2.3.x, 2.4.x, and 2.5.x."

words/protocols/jabber/sasl.py, line 101, calls base64.b64encode, which doesn't exist in Python 2.3.x. 

Replacing b64encode with the base64.encodestring seems to fix the problem.

<details><summary>Searchable metadata</summary>

```
trac-id__2461 2461
type__defect defect
reporter__murphy_slaw murphy_slaw
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1171920876000000 1171920876000000
changetime__1173453598000000 1173453598000000
version__None None
owner__ralphm ralphm
cc__murphy_slaw cc__ralphm cc__itamarst cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst set owner to @ralphm|
|-|-|

Since the buildbot didn't catch this, there would appear to be a lack in test coverage as well.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm set status to `assigned`|
|-|-|

I've looked briefly into this. The `encodestring` from `base64` behaves different in that it adds a `\n` to the end of the result. I suppose that can be worked around.

Started a branch to work on this with some preliminary tests: [branches/jabber-sasl-base64-2461.](branches/jabber-sasl-base64-2461.)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm set owner to @itamarst<br>@ralphm set status to `new`|
|-|-|

Added more tests and fix in the mentioned branch. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @ralphm|
|-|-|

 * Unused import sasl_mechanisms from test_jabbersasl. It probably comes from the commented implements
 * the raw 'except:' in onChallenge could probably be changed to something more specific
 * the raise should use an instance (raise SASLIncorrectEncoding(), raise SASLNoAcceptableMechanism())
 * copyright 2001-2007 in test_jabbersasl
 * some missing docstrings in test_jabbersasl
 * trailing whitespaces in test_jabbersasl

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to @therve|
|-|-|

Thanks for the comments. All addressed. Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @ralphm|
|-|-|

 * There still undocumented parts in test_jabbersasl (eg SASLInitiatingInitializerTest)
 * setMechanism of SASLInitiatingInitializer is not tested
 * sendResponse with data is also not tested

As this branch is already a big improve on the current situation, I don't think it should block you from merging, though.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set status to `closed`|
|-|-|

(In [19791]) Unbreak XMPP SASL on Python 2.3.

Author: ralphm
Reviewer: therve
Fixes [#2461](https://github.com/twisted/twisted/issues/2461).

Use the legacy interface to base64 for Python 2.3, and also implement some
additional checks for inappropriate inputs before actually decoding
challenges.
